internal/controllers/machinedeployment/machinedeployment_status.go

```diff
@@ -82,11 +93,41 @@ func (r *Reconciler) updateStatus(ctx context.Context, s *scope) (retErr error)
 // as a consequence it is required to compute the counters again before calling scale down machine sets,
 // and again to before computing the overall availability of the Machine deployment.
 func setReplicas(machineDeployment *clusterv1.MachineDeployment, machineSets []*clusterv1.MachineSet) {
+	machineDeployment.Status.Replicas = mdutil.GetActualReplicaCountForMachineSets(machineSets)
 	machineDeployment.Status.ReadyReplicas = mdutil.GetReadyReplicaCountForMachineSets(machineSets)
 	machineDeployment.Status.AvailableReplicas = mdutil.GetAvailableReplicaCountForMachineSets(machineSets)
 	machineDeployment.Status.UpToDateReplicas = mdutil.GetUptoDateReplicaCountForMachineSets(machineSets)
 }
 
+func setPhase(_ context.Context, machineDeployment *clusterv1.MachineDeployment, machineSets []*clusterv1.MachineSet, getAndAdoptMachineSetsForDeploymentSucceeded bool) {
+	if !getAndAdoptMachineSetsForDeploymentSucceeded || machineDeployment.Spec.Replicas == nil {
+		machineDeployment.Status.Phase = string(clusterv1.MachineDeploymentPhaseUnknown)
+		return
+	}
+
+	desiredReplicas := ptr.Deref(machineDeployment.Spec.Replicas, 0)
+	if !machineDeployment.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+	currentReplicas := mdutil.GetActualReplicaCountForMachineSets(machineSets)
+
+	if desiredReplicas < currentReplicas {
+		machineDeployment.Status.Phase = string(clusterv1.MachineDeploymentPhaseScalingDown)
+		return
+	}
+	if desiredReplicas > currentReplicas {
+		machineDeployment.Status.Phase = string(clusterv1.MachineDeploymentPhaseScalingUp)
+		return
+	}
+	if desiredReplicas == currentReplicas {
+		machineDeployment.Status.Phase = string(clusterv1.MachineDeploymentPhaseRunning)
+		return
+	}
+
+	// NOTE: this should never happen.
```

Comment:
I would suggest to move the 3 if's above into a switch case with the last case as the default case. It's safe to do and we don't need this handling here (and we don't need the returns)